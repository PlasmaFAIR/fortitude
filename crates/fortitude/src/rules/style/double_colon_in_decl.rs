use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What does it do?
/// Checks for missing double-colon separator in variable declarations.
///
/// ## Why is this bad?
/// The double-colon separator is required when declaring variables with
/// attributes, so for consistency, all variable declarations should use it.
#[derive(ViolationMetadata)]
pub(crate) struct MissingDoubleColon {}

impl AlwaysFixableViolation for MissingDoubleColon {
    #[derive_message_formats]
    fn message(&self) -> String {
        "variable declaration missing '::'".to_string()
    }

    fn fix_title(&self) -> String {
        "Add '::'".to_string()
    }
}
impl AstRule for MissingDoubleColon {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if node
            .children(&mut node.walk())
            .filter_map(|child| child.to_text(src.source_text()))
            .all(|child| child != "::")
        {
            let first_decl = node.child_by_field_name("declarator")?;
            let start_pos = first_decl.start_textsize();
            let fix = Fix::safe_edit(Edit::insertion(":: ".to_string(), start_pos));
            some_vec!(Diagnostic::from_node(Self {}, node).with_fix(fix))
        } else {
            None
        }
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["variable_declaration"]
    }
}
