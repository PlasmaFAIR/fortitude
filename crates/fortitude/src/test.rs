use std::path::Path;

use anyhow::Result;
use ruff_source_file::{SourceFile, SourceFileBuilder};

use crate::{
    check::{
        ast_entrypoint_map, check_file, read_to_string, rules_to_path_rules, rules_to_text_rules,
    },
    message::{Emitter, TextEmitter},
    rule_table::RuleTable,
    rules::Rule,
    settings::{self, FixMode, Settings},
};

#[macro_export]
macro_rules! apply_common_filters {
    {} => {
        let mut settings = insta::Settings::clone_current();
        // Convert windows paths to Unix Paths.
        settings.add_filter(r"\\\\?([\w\d.])", "/$1");
        let _bound = settings.bind_to_scope();
    }
}

pub(crate) fn test_resource_path(path: impl AsRef<Path>) -> std::path::PathBuf {
    Path::new("./resources/test/").join(path)
}

/// Run [`check_path`] on a Fortran file in the `resources/test/fixtures` directory.
pub(crate) fn test_path(
    path: impl AsRef<Path>,
    rules: &[Rule],
    settings: &Settings,
) -> Result<String> {
    let path = test_resource_path("fixtures").join(path);
    let source = read_to_string(&path)?;
    let filename = path.to_string_lossy();
    let file = SourceFileBuilder::new(filename.as_ref(), source.as_str()).finish();
    Ok(test_contents(&path, &file, rules, settings))
}

pub(crate) fn test_contents(
    path: &Path,
    file: &SourceFile,
    rules: &[Rule],
    settings: &Settings,
) -> String {
    let rule_table = RuleTable::from_iter(rules.iter().cloned());
    let path_rules = rules_to_path_rules(&rule_table);
    let text_rules = rules_to_text_rules(&rule_table);
    let ast_entrypoints = ast_entrypoint_map(&rule_table);

    match check_file(
        &rule_table,
        &path_rules,
        &text_rules,
        &ast_entrypoints,
        path,
        file,
        settings,
        FixMode::Generate,
        settings::IgnoreAllowComments::Disabled,
    ) {
        Ok(violations) => {
            if violations.messages.is_empty() {
                return String::new();
            }

            let mut output = Vec::new();

            TextEmitter::default()
                .with_show_fix_status(true)
                .with_show_fix_diff(true)
                .with_show_source(true)
                .with_unsafe_fixes(crate::settings::UnsafeFixes::Enabled)
                .emit(&mut output, &violations.messages)
                .unwrap();

            String::from_utf8(output).unwrap()
        }
        Err(msg) => {
            panic!("Failed to process: {msg}");
        }
    }
}
