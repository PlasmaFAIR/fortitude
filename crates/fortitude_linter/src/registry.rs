use fortitude_macros::RuleNamespace;
use rayon::str;
use std::str::FromStr; // Needed by strum_macros

pub use crate::rules::Rule;
pub use rule_set::{RuleSet, RuleSetIterator};

mod rule_set;

// Rule categories and identity codes
// ----------------------------------
// Helps to sort rules into logical categories, and acts as a unique tag with which
// users can switch rules on and off.

pub trait AsRule {
    fn rule(&self) -> Rule;
}

impl Rule {
    pub fn from_code(code: &str) -> Result<Self, FromCodeError> {
        let (category, code) = Category::parse_code(code).ok_or(FromCodeError::Unknown)?;
        category
            .all_rules()
            .find(|rule| rule.noqa_code().suffix() == code)
            .ok_or(FromCodeError::Unknown)
    }
}

#[derive(thiserror::Error, Debug)]
pub enum FromCodeError {
    #[error("unknown rule code")]
    Unknown,
}

/// The category of each rule defines the sort of problem it intends to solve.
#[derive(
    Debug,
    PartialEq,
    Eq,
    Clone,
    Hash,
    PartialOrd,
    Ord,
    strum_macros::AsRefStr,
    strum_macros::Display,
    strum_macros::EnumIter,
    strum_macros::EnumString,
    strum_macros::IntoStaticStr,
    RuleNamespace,
)]
#[repr(u16)]
#[strum(serialize_all = "kebab-case")]
pub enum Category {
    /// Failure to parse a file.
    #[prefix = "E"]
    Error,
    /// Detect code that is bug-prone or likely to be incorrect.
    #[prefix = "C"]
    Correctness,
    /// Obsolescent language features, as determined by the Fortran standard.
    #[prefix = "OB"]
    Obsolescent,
    /// Update to modern Fortran features. Used for less severe issues than `Obsolescent`,
    /// and goes beyond recommendations in the Fortran standard.
    #[prefix = "MOD"]
    Modernisation,
    /// Violation of style conventions.
    #[prefix = "S"]
    Style,
    /// Avoid platform/compiler-specific features.
    #[prefix = "PORT"]
    Portability,
    /// Fortitude specific rules.
    #[prefix = "FORT"]
    Fortitude,
}

pub trait RuleNamespace: Sized {
    /// Returns the prefix that every single code that fortitude uses to identify
    /// rules from this category starts with.
    fn common_prefix(&self) -> &'static str;

    /// Returns the variant corresponding to the given prefix.
    fn from_prefix(preifx: &str) -> Result<Self, &'static str>;

    /// Attempts to parse the given rule code. If the prefix is recognized
    /// returns the respective variant along with the code with the common
    /// prefix stripped.
    fn parse_code(code: &str) -> Option<(Self, &str)>;

    #[allow(dead_code)]
    fn name(&self) -> &'static str;

    #[allow(dead_code)]
    fn description(&self) -> &'static str;
}

pub mod clap_completion {
    use clap::builder::{PossibleValue, TypedValueParser, ValueParserFactory};
    use strum::IntoEnumIterator;

    use crate::registry::Rule;

    #[derive(Clone)]
    pub struct RuleParser;

    impl ValueParserFactory for Rule {
        type Parser = RuleParser;

        fn value_parser() -> Self::Parser {
            RuleParser
        }
    }

    impl TypedValueParser for RuleParser {
        type Value = Rule;

        fn parse_ref(
            &self,
            cmd: &clap::Command,
            arg: Option<&clap::Arg>,
            value: &std::ffi::OsStr,
        ) -> Result<Self::Value, clap::Error> {
            let value = value
                .to_str()
                .ok_or_else(|| clap::Error::new(clap::error::ErrorKind::InvalidUtf8))?;

            Rule::from_code(value).map_err(|_| {
                let mut error =
                    clap::Error::new(clap::error::ErrorKind::ValueValidation).with_cmd(cmd);
                if let Some(arg) = arg {
                    error.insert(
                        clap::error::ContextKind::InvalidArg,
                        clap::error::ContextValue::String(arg.to_string()),
                    );
                }
                error.insert(
                    clap::error::ContextKind::InvalidValue,
                    clap::error::ContextValue::String(value.to_string()),
                );
                error
            })
        }

        fn possible_values(&self) -> Option<Box<dyn Iterator<Item = PossibleValue> + '_>> {
            Some(Box::new(Rule::iter().map(|rule| {
                let name = rule.noqa_code().to_string();
                let help = rule.as_ref().to_string();
                PossibleValue::new(name).help(help)
            })))
        }
    }
}

#[cfg(test)]
mod tests {
    use std::mem::size_of;

    use strum::IntoEnumIterator;

    use super::{Category, Rule, RuleNamespace};

    #[test]
    fn documentation() {
        for rule in Rule::iter() {
            assert!(
                rule.explanation().is_some(),
                "Rule {} is missing documentation",
                rule.as_ref()
            );
        }
    }

    #[test]
    fn check_code_serialization() {
        for rule in Rule::iter() {
            assert!(
                Rule::from_code(&format!("{}", rule.noqa_code())).is_ok(),
                "{rule:?} could not be round-trip serialized."
            );
        }
    }

    #[test]
    fn category_parse_code() {
        for rule in Rule::iter() {
            let code = format!("{}", rule.noqa_code());
            let (category, rest) =
                Category::parse_code(&code).unwrap_or_else(|| panic!("couldn't parse {code:?}"));
            assert_eq!(code, format!("{}{rest}", category.common_prefix()));
        }
    }

    #[test]
    fn rule_size() {
        assert_eq!(2, size_of::<Rule>());
    }
}
