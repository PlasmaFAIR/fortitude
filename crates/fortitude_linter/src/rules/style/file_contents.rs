use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for multiple modules in one file
///
/// ## Why is this bad?
/// Placing each module into its own file improves maintainability
/// by making each module easier to locate for developers, and also
/// making dependency generation in build systems easier.
#[derive(ViolationMetadata)]
pub(crate) struct MultipleModules {}

impl Violation for MultipleModules {
    #[derive_message_formats]
    fn message(&self) -> String {
        "Multiple modules in one file, split into one module per file".to_string()
    }
}

impl AstRule for MultipleModules {
    fn check(_settings: &Settings, node: &Node, _src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let violations: Vec<Diagnostic> = node
            .children(&mut node.walk())
            .filter(|node| node.kind() == "module")
            .skip(1)
            .map(|m| -> Diagnostic {
                let m_first = m.child(0).unwrap_or(m);
                Diagnostic::from_node(MultipleModules {}, &m_first)
            })
            .collect();

        Some(violations)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["translation_unit"]
    }
}

/// ## What it does
/// Checks for programs and modules in one file
///
/// ## Why is this bad?
/// Separating top-level constructs into their own files improves
/// maintainability by making each easier to locate for developers,
/// and also making dependency generation in build systems easier.
#[derive(ViolationMetadata)]
pub(crate) struct ProgramWithModule {}

impl Violation for ProgramWithModule {
    #[derive_message_formats]
    fn message(&self) -> String {
        "Program and module in one file, split into their own files".to_string()
    }
}

impl AstRule for ProgramWithModule {
    fn check(_settings: &Settings, node: &Node, _src: &SourceFile) -> Option<Vec<Diagnostic>> {
        // There must be a program statement to trigger this rule
        if !node
            .children(&mut node.walk())
            .any(|node| node.kind() == "program")
        {
            return None;
        }

        // Mark the violation on the second, and subsequent, occurrences
        let violations: Vec<Diagnostic> = node
            .children(&mut node.walk())
            .filter(|node| node.kind() == "module" || node.kind() == "program")
            .skip(1)
            .map(|m| -> Diagnostic {
                let m_first = m.child(0).unwrap_or(m);
                Diagnostic::from_node(ProgramWithModule {}, &m_first)
            })
            .collect();

        Some(violations)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["translation_unit"]
    }
}
