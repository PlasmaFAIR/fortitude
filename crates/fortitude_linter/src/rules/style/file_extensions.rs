use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_text_size::TextRange;

use crate::settings::Settings;
use crate::PathRule;
use std::path::Path;

/// ## What it does
/// Checks for use of standard file extensions.
///
/// ## Why is it bad?
/// The standard file extensions for modern (free-form) Fortran are '.f90' or  '.F90'.
/// Forms that reference later Fortran standards such as '.f08' or '.F95' may be rejected
/// by some compilers and build tools.
#[derive(ViolationMetadata)]
pub(crate) struct NonStandardFileExtension {}

impl Violation for NonStandardFileExtension {
    #[derive_message_formats]
    fn message(&self) -> String {
        "file extension should be '.f90' or '.F90'".to_string()
    }
}

impl PathRule for NonStandardFileExtension {
    fn check(_settings: &Settings, path: &Path) -> Option<Diagnostic> {
        match path.extension() {
            Some(ext) => {
                // Must check like this as ext is an OsStr
                if ["f90", "F90"].iter().any(|&x| x == ext) {
                    None
                } else {
                    Some(Diagnostic::new(Self {}, TextRange::default()))
                }
            }
            None => Some(Diagnostic::new(Self {}, TextRange::default())),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bad_file_extension() {
        let path = Path::new("my/dir/to/file.f95");
        assert_eq!(
            NonStandardFileExtension::check(&Settings::default(), path),
            Some(Diagnostic::new(
                NonStandardFileExtension {},
                TextRange::default()
            )),
        );
    }

    #[test]
    fn test_missing_file_extension() {
        let path = Path::new("my/dir/to/file");
        assert_eq!(
            NonStandardFileExtension::check(&Settings::default(), path),
            Some(Diagnostic::new(
                NonStandardFileExtension {},
                TextRange::default()
            )),
        );
    }

    #[test]
    fn test_correct_file_extensions() {
        let path1 = Path::new("my/dir/to/file.f90");
        let path2 = Path::new("my/dir/to/file.F90");
        assert_eq!(
            NonStandardFileExtension::check(&Settings::default(), path1),
            None
        );
        assert_eq!(
            NonStandardFileExtension::check(&Settings::default(), path2),
            None
        );
    }
}
