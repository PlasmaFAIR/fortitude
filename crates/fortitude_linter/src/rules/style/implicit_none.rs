/// Defines rules that raise errors if implicit typing is in use.
use crate::ast::FortitudeNode;
use crate::rules::correctness::implicit_typing::{
    child_is_implicit_none, implicit_statement_is_none,
};
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Fix};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for unnecessary `implicit none` in module procedures
///
/// ## Why is this bad?
/// If a module has 'implicit none' set, it is not necessary to set it in contained
/// functions and subroutines (except when using interfaces).
#[derive(ViolationMetadata)]
pub struct SuperfluousImplicitNone {
    entity: String,
}

impl AlwaysFixableViolation for SuperfluousImplicitNone {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { entity } = self;
        format!("'implicit none' set on the enclosing {entity}")
    }

    fn fix_title(&self) -> String {
        "Remove unnecessary 'implicit none'".to_string()
    }
}

impl AstRule for SuperfluousImplicitNone {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if !implicit_statement_is_none(node) {
            return None;
        }
        let parent = node.parent()?;
        if matches!(parent.kind(), "function" | "subroutine") {
            for ancestor in parent.ancestors() {
                let kind = ancestor.kind();
                match kind {
                    "module" | "submodule" | "program" | "function" | "subroutine" => {
                        if !child_is_implicit_none(&ancestor) {
                            continue;
                        }
                        let entity = kind.to_string();
                        let fix = Fix::safe_edit(node.edit_delete(src));
                        return some_vec![
                            Diagnostic::from_node(Self { entity }, node).with_fix(fix)
                        ];
                    }
                    "interface" => {
                        break;
                    }
                    _ => {
                        continue;
                    }
                }
            }
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["implicit_statement"]
    }
}
