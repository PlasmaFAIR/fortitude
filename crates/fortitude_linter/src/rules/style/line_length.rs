/// Defines rules that govern line length.
use crate::settings::Settings;
use crate::TextRule;
use lazy_regex::regex_is_match;
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use ruff_source_file::UniversalNewlines;
use ruff_text_size::{TextLen, TextRange, TextSize};

/// ## What does it do?
/// Checks line length isn't too long
///
/// ## Why is this bad?
/// Long lines are more difficult to read, and may not fit on some developers'
/// terminals. The line continuation character '&' may be used to split a long line
/// across multiple lines, and overly long expressions may be broken down into
/// multiple parts.
///
/// The maximum line length can be changed using the flag `--line-length=N`. The
/// default maximum line length is 100 characters. This is a fair bit more than the
/// traditional 80, but due to the verbosity of modern Fortran it can sometimes be
/// difficult to squeeze lines into that width, especially when using large indents
/// and multiple levels of indentation.
///
/// Some lines that are longer than the maximum length may be acceptable, such as
/// long strings or comments. This is to allow for long URLs or other text that cannot
/// be reasonably split across multiple lines.
///
/// Note that the Fortran standard states a maximum line length of 132 characters,
/// and while some modern compilers will support longer lines, for portability it
/// is recommended to stay beneath this limit.
#[derive(ViolationMetadata)]
pub(crate) struct LineTooLong {
    max_length: usize,
    actual_length: usize,
}

impl Violation for LineTooLong {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self {
            max_length,
            actual_length,
        } = self;
        format!("line length of {actual_length}, exceeds maximum {max_length}")
    }
}

impl TextRule for LineTooLong {
    fn check(settings: &Settings, source_file: &SourceFile) -> Vec<Diagnostic> {
        let source = source_file.to_source_code();
        let max_length = settings.check.line_length;
        let mut violations = Vec::new();
        for line in source.text().universal_newlines() {
            // Note: Can't use string.len(), as that gives byte length, not char length
            let actual_length = line.chars().count();
            if actual_length > max_length {
                // Are we ending on a string or comment? If so, we'll allow it through, as it may
                // contain something like a long URL that cannot be reasonably split across multiple
                // lines.
                if regex_is_match!(r#"(["']\w*&?$)|(!.*$)|(^\w*&)"#, line.as_str()) {
                    continue;
                }
                // Get the byte range from the first character that oversteps the limit
                // to the end of the line
                let extra_bytes: TextSize = line
                    .chars()
                    .rev()
                    .take(actual_length - max_length)
                    .map(TextLen::text_len)
                    .sum();
                let range = TextRange::new(line.end() - extra_bytes, line.end());
                violations.push(Diagnostic::new(
                    Self {
                        max_length,
                        actual_length,
                    },
                    range,
                ));
            }
        }
        violations
    }
}
