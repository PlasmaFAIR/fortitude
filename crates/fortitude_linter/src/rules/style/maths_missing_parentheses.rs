use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Fix};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks if mathematical expressions are missing parentheses when operators have
/// different precedences.
///
/// ## Why is this bad?
/// Long or complex expressions can be difficult or confusing to read, especially when
/// mixing operators with different precedences. Adding parentheses can clarify the code
/// and make the author's intent clearer, reducing the likelihood of misunderstandings
/// or bugs.
///
/// ## Example
/// ```f90
/// x = 1. + 2. * 3. - 4. / 5.
/// ```
///
/// Use instead:
/// ```f90
/// x = 1. + (2. * 3.) - (4. / 5.)
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct MathsMissingParentheses {
    op1: String,
    op2: String,
}

impl AlwaysFixableViolation for MathsMissingParentheses {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { op1, op2 } = &self;
        format!("'{op1}' has higher precedence than '{op2}'; add parentheses for clarity")
    }

    fn fix_title(&self) -> String {
        "Add parentheses".to_string()
    }
}

#[derive(PartialEq)]
enum Operator {
    Exponential,
    Multiplication,
    Division,
    Addition,
    Subtraction,
}

impl Operator {
    fn precedence(&self) -> u8 {
        use Operator::*;
        match self {
            Exponential => 4,
            Multiplication | Division => 3,
            Addition | Subtraction => 2,
        }
    }
}

impl TryFrom<&str> for Operator {
    type Error = &'static str;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        use Operator::*;
        match value {
            "**" => Ok(Exponential),
            "*" => Ok(Multiplication),
            "/" => Ok(Division),
            "-" => Ok(Addition),
            "+" => Ok(Subtraction),
            _ => Err("not a maths operator"),
        }
    }
}

impl From<Operator> for String {
    fn from(value: Operator) -> String {
        use Operator::*;
        match value {
            Exponential => "**",
            Multiplication => "*",
            Division => "/",
            Addition => "-",
            Subtraction => "+",
        }
        .to_string()
    }
}

impl AstRule for MathsMissingParentheses {
    fn check<'a>(
        _settings: &Settings,
        node: &'a Node,
        src: &'a SourceFile,
    ) -> Option<Vec<Diagnostic>> {
        let text = src.source_text();
        let op: Operator = node.child(1)?.to_text(text)?.try_into().ok()?;

        let parent = node.parent()?;
        if !matches!(parent.kind(), "math_expression" | "unary_expression") {
            return None;
        }

        let edit = node.edit_replacement(src, format!("({})", node.to_text(text)?));
        let fix = Fix::safe_edit(edit);
        if op == Operator::Exponential && parent.kind() == "unary_expression" {
            let op2 = parent.child(0)?.to_text(text)?.to_string();
            let op1: String = op.into();
            return some_vec!(Diagnostic::from_node(Self { op1, op2 }, node).with_fix(fix));
        }

        let parent_op: Operator = parent.child(1)?.to_text(text)?.try_into().ok()?;

        if op.precedence() == parent_op.precedence() {
            return None;
        }

        let (op1, op2): (String, String) = if op.precedence() > parent_op.precedence() {
            (op.into(), parent_op.into())
        } else {
            (parent_op.into(), op.into())
        };

        some_vec!(Diagnostic::from_node(Self { op1, op2 }, node).with_fix(fix))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["math_expression"]
    }
}
