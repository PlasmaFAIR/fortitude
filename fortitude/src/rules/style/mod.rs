pub mod double_colon_in_decl;
pub mod end_statements;
pub mod file_contents;
pub mod file_extensions;
pub mod functions;
pub mod implicit_none;
pub mod keywords;
pub mod line_length;
pub mod semicolons;
pub mod whitespace;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::rules::style::keywords;
    use crate::settings::{CheckSettings, Settings};
    use crate::test::test_path;

    #[test_case(Rule::LineTooLong, Path::new("S001.f90"))]
    #[test_case(Rule::UnnamedEndStatement, Path::new("S061.f90"))]
    #[test_case(Rule::MissingDoubleColon, Path::new("S071.f90"))]
    #[test_case(Rule::SuperfluousSemicolon, Path::new("S081.f90"))]
    #[test_case(Rule::MultipleStatementsPerLine, Path::new("S082.f90"))]
    #[test_case(Rule::TrailingWhitespace, Path::new("S101.f90"))]
    #[test_case(Rule::IncorrectSpaceBeforeComment, Path::new("S102.f90"))]
    #[test_case(Rule::SuperfluousImplicitNone, Path::new("S201.f90"))]
    #[test_case(Rule::MultipleModules, Path::new("S211.f90"))]
    #[test_case(Rule::ProgramWithModule, Path::new("S212.f90"))]
    #[test_case(Rule::FunctionMissingResult, Path::new("S221.f90"))]
    #[test_case(Rule::KeywordsMissingSpace, Path::new("S231.f90"))]
    #[test_case(Rule::KeywordHasWhitespace, Path::new("S231.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::SuperfluousImplicitNone, Path::new("S201_ok.f90"))]
    fn rules_pass(rule_code: Rule, path: &Path) -> Result<()> {
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        assert!(
            diagnostics.is_empty(),
            "Test source has no warnings, but some were raised:\n{diagnostics}"
        );
        Ok(())
    }

    #[test_case(Rule::LineTooLong, Path::new("S001_line_length_20.f90"))]
    fn line_too_long_line_length_20(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());

        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                line_length: 20,
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::KeywordsMissingSpace, Path::new("S231.f90"))]
    #[test_case(Rule::KeywordHasWhitespace, Path::new("S231.f90"))]
    fn keyword_whitespace_include_inout_goto(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!(
            "{}_{}_include_inout_goto",
            rule_code.as_ref(),
            path.to_string_lossy()
        );

        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                keyword_whitespace: keywords::settings::Settings {
                    inout_with_space: true,
                    goto_with_space: true,
                },
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
