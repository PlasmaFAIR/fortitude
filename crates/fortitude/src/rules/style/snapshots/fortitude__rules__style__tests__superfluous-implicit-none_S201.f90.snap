---
source: fortitude/src/rules/style/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/style/S201.f90:5:5: S201 [*] 'implicit none' set on the enclosing module
  |
3 | contains
4 |   integer function myfunc(x)
5 |     implicit none
  |     ^^^^^^^^^^^^^ S201
6 |     integer, intent(in) :: x
7 |     myfunc = x * 2
  |
  = help: Remove unnecessary 'implicit none'

ℹ Safe fix
2 2 |   implicit none
3 3 | contains
4 4 |   integer function myfunc(x)
5   |-    implicit none
6 5 |     integer, intent(in) :: x
7 6 |     myfunc = x * 2
8 7 |   end function myfunc

./resources/test/fixtures/style/S201.f90:10:5: S201 [*] 'implicit none' set on the enclosing module
   |
 8 |   end function myfunc
 9 |   subroutine mysub(x)
10 |     implicit none
   |     ^^^^^^^^^^^^^ S201
11 |     integer, intent(inout) :: x
12 |     x = x * 2
   |
   = help: Remove unnecessary 'implicit none'

ℹ Safe fix
7  7  |     myfunc = x * 2
8  8  |   end function myfunc
9  9  |   subroutine mysub(x)
10    |-    implicit none
11 10 |     integer, intent(inout) :: x
12 11 |     x = x * 2
13 12 |   end subroutine mysub

./resources/test/fixtures/style/S201.f90:23:5: S201 [*] 'implicit none' set on the enclosing program
   |
21 | contains
22 |   integer function myfunc2(x)
23 |     implicit none
   |     ^^^^^^^^^^^^^ S201
24 |     integer, intent(in) :: x
25 |     myfunc2 = x * 2
   |
   = help: Remove unnecessary 'implicit none'

ℹ Safe fix
20 20 | 
21 21 | contains
22 22 |   integer function myfunc2(x)
23    |-    implicit none
24 23 |     integer, intent(in) :: x
25 24 |     myfunc2 = x * 2
26 25 |   end function myfunc2

./resources/test/fixtures/style/S201.f90:28:5: S201 [*] 'implicit none' set on the enclosing program
   |
26 |   end function myfunc2
27 |   subroutine mysub2(x)
28 |     implicit none
   |     ^^^^^^^^^^^^^ S201
29 |     integer, intent(inout) :: x
30 |     x = x * 2
   |
   = help: Remove unnecessary 'implicit none'

ℹ Safe fix
25 25 |     myfunc2 = x * 2
26 26 |   end function myfunc2
27 27 |   subroutine mysub2(x)
28    |-    implicit none
29 28 |     integer, intent(inout) :: x
30 29 |     x = x * 2
31 30 |   end subroutine mysub2
